pkg/component/operator/web/v0/crawl_website.go

```diff
@@ -198,6 +209,33 @@ func (e *execution) CrawlWebsite(input *structpb.Struct) (*structpb.Struct, erro
 		c.Wait()
 	}()
 
+	// To avoid to wait for 2 minutes, we use a timer to check if there is a new page.
+	// If there is no new page, we cancel the context.
+	go func() {
+		inactivityTimer := time.NewTimer(2 * time.Second)
+		defer inactivityTimer.Stop()
+
+		for {
+			select {
+			case <-pageUpdateCh:
+				// Reset the timer whenever we get a new page
+				if !inactivityTimer.Stop() {
+					select {
+					case <-inactivityTimer.C:
+					default:
+					}
+				}
+				inactivityTimer.Reset(2 * time.Second)
+			case <-inactivityTimer.C:
+				// If no new pages for 2 seconds, cancel the context
+				cancel()
+				return
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
 	<-ctx.Done()
```

Comment:
Is this redundant with the previous `<-ctx.Done()`?

Reply:
We need this to block the main execution.

Reply:
Yeah but you can skip the `go func()` in the previous block and run the `for` / `select` in the main thread, right? Otherwise it looks like you're dispatching work in order to wait for it immediately.

Reply:
Oh, do you mean we actually don't need another go-routine for this block?
So, `<-ctx.Done()` in the main thread is not necessary.

```
		inactivityTimer := time.NewTimer(2 * time.Second)
		defer inactivityTimer.Stop()

		for {
			select {
			case <-pageUpdateCh:
				inactivityTimer.Reset(2 * time.Second)
			// If no new pages for 2 seconds, cancel the context
			case <-inactivityTimer.C:
				cancel()
				return
			// If the context is done, we should return
			case <-ctx.Done():
				return
			}
		}
```

Reply:
> depending on the thread about c.Wait()

It does not work, so I modified the code [here](https://github.com/instill-ai/pipeline-backend/pull/819/commits/4a9055a3dfd9a03ec24da8527428bebf70f06a74).